File: engine/tests/fixtures/comprehensive.swift
import Foundation

enum Mode {
    case fast
    case safe
}

protocol Solvable {
    func solve(_ limit: Int) -> [Int: Int]
}

final class Solver: Solvable {
    static let version = 1
    let values: [Int]

    init(values: [Int]) {
        self.values = values
    }

    func solve(_ limit: Int = 3) -> [Int: Int] {
        var seen: [Int: Int] = [:]
        let unique = Set(values)
        let prefix = Array(values[0..<limit])
        for value in prefix {
            if value > 0 {
                seen[value] = solve(limit - 1).count
            }
        }
        var remaining = limit
        while remaining > 0 {
            remaining -= 1
        }
        return seen
    }
}

File: engine/tests/fixtures/comprehensive.go
package algorithms

import "fmt"

type Mode int

type Solvable interface {
    Solve(limit int) map[int]int
}

const (
    Fast Mode = iota
    Safe
)

type Solver struct {
    values []int
}

func NewSolver(values []int) *Solver {
    return &Solver{values: values}
}

func (solver *Solver) Solve(limit int) map[int]int {
    seen := make(map[int]int)
    unique := make(map[int]struct{})
    prefix := solver.values[:limit]
    for _, value := range prefix {
        if value > 0 {
            seen[value] = len(solver.Solve(limit - 1))
        }
    }
    for limit > 0 {
        limit -= 1
    }
    fmt.Println(unique)
    return seen
}
